### PR TITLE
tsp, enable sync stack

### PR DESCRIPTION
**Pull request by XiaofeiCao**
issue: https://github.com/Azure/autorest.java/issues/2038

**Use case test for sync stack:**
```typescript
@route("/syncstack")
interface SyncStackOp {
  // get
  get is ResourceOperations.ResourceRead<Resource>;
  
  // pageable
  list is ResourceOperations.ResourceList<Resource>;

  // lro
  @pollingOperation(ResourceOperations.GetResourceOperationStatus<Resource, never>)
  createOrUpdate is ResourceOperations.ResourceCreateOrUpdate<Resource>;

  @pollingOperation(ResourceOperations.GetResourceOperationStatus<Resource, never>)
  delete is ResourceOperations.LongRunningResourceDelete<Resource>;

  @pollingOperation(ResourceOperations.GetResourceOperationStatus<Resource, ExportedResource>)
  export is ResourceOperations.LongRunningResourceAction<Resource, ExportParams, ExportedResource>;
}
```

**Generated protocol methods:**
```java
@Generated
@ServiceMethod(returns = ReturnType.SINGLE)
public Response<BinaryData> getWithResponse(String name, RequestOptions requestOptions) {
    return this.serviceClient.getWithResponse(name, requestOptions);
}

@Generated
@ServiceMethod(returns = ReturnType.COLLECTION)
public PagedIterable<BinaryData> list(RequestOptions requestOptions) {
    return this.serviceClient.list(requestOptions);
}

@Generated
@ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
public SyncPoller<BinaryData, BinaryData> beginCreateOrUpdate(
        String name, BinaryData resource, RequestOptions requestOptions) {
    return this.serviceClient.beginCreateOrUpdate(name, resource, requestOptions);
}

@Generated
@ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
public SyncPoller<BinaryData, Void> beginDelete(String name, RequestOptions requestOptions) {
    return this.serviceClient.beginDelete(name, requestOptions);
}

@Generated
@ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
public SyncPoller<BinaryData, BinaryData> beginExport(
        String name, String projectFileVersion, RequestOptions requestOptions) {
    return this.serviceClient.beginExport(name, projectFileVersion, requestOptions);
}
```

**Generated convenience methods:**
```java
@Generated
@ServiceMethod(returns = ReturnType.SINGLE)
public Resource get(String name) {
    // Generated convenience method for getWithResponse
    RequestOptions requestOptions = new RequestOptions();
    return getWithResponse(name, requestOptions).getValue().toObject(Resource.class);
}

@Generated
@ServiceMethod(returns = ReturnType.COLLECTION)
public PagedIterable<Resource> list() {
    // Generated convenience method for list
    RequestOptions requestOptions = new RequestOptions();
    return serviceClient.list(requestOptions).mapPage(value -> value.toObject(Resource.class));
}

@Generated
@ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
public SyncPoller<ResourceOperationStatusResourceError, Void> beginDelete(String name) {
    // Generated convenience method for beginDeleteWithModel
    RequestOptions requestOptions = new RequestOptions();
    return serviceClient.beginDeleteWithModel(name, requestOptions);
}

@Generated
@ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
public SyncPoller<
                ResourceOperationStatusResourceExportedResourceError,
                ResourceOperationStatusResourceExportedResourceError>
        beginExport(String name, String projectFileVersion) {
    // Generated convenience method for beginExportWithModel
    RequestOptions requestOptions = new RequestOptions();
    return serviceClient.beginExportWithModel(name, projectFileVersion, requestOptions);
}
```

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java [hunk 1]
```diff
@@ -149,6 +123,28 @@ public static void getAsyncSyncClients(Client client, ServiceClient serviceClien
...
+                            .filter(m -> m.getMethodVisibility() == JavaVisibility.Public)
+                            .collect(Collectors.toList());
+                    if (!cMethods.isEmpty()) {
+                        // sync stack generates additional proxy methods with name suffix "Sync"
+                        Set<String> proxyMethodNames = cMethods.stream()
+                                .map(m -> m.getProxyMethod().getName())
+                                .collect(Collectors.toSet());
+                        return serviceClient.getClientMethods().stream()
+                                .filter(m -> proxyMethodNames.contains(m.getProxyMethod().getName()) && m.getMethodVisibility() == JavaVisibility.Public)
+                                .map(m -> new ConvenienceMethod(m, cMethods));
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:141]: Real change here. Others in this class are refactors.

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java [hunk 2]
```diff
@@ -149,6 +125,28 @@ public static void getAsyncSyncClients(Client client, ServiceClient serviceClien
...
+                            .stream()
+                            .filter(m -> m.getMethodVisibility() == JavaVisibility.Public)
+                            .collect(Collectors.toList());
+                    if (!cMethods.isEmpty()) {
+                        // sync stack generates additional proxy methods with name suffix "Sync"
+                        Set<String> proxyMethodNames = cMethods.stream()
+                                .map(m -> m.getProxyMethod().getName())
+                                .collect(Collectors.toSet());
+                        return clientMethods.get().stream()
+                                .filter(m -> proxyMethodNames.contains(m.getProxyMethod().getName()) && m.getMethodVisibility() == JavaVisibility.Public)
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:142]: I am not sure "contains" is best solution. Probably more specific about "Sync"?

E.g. client may have "list" and "listByFoo", or something like that.

File: typespec-tests/tsp/sync-stack.tsp
Comment (weidongxu-microsoft) [inline typespec-tests/tsp/sync-stack.tsp:1]: not sure we really need it?

Comment (XiaofeiCao) [reply typespec-tests/tsp/sync-stack.tsp:1]: Initially I added it to test common use cases are generated as expected. 

But yeah, I think they are already covered in other tests. Will remove it and add one test for sync pageable in `CoreTests` and call it a day.

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java (hunk 2 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java]: Got it. Replaced with proxyMethod.getBaseName(). Thanks!

**Review by weidongxu-microsoft — APPROVED** [review]
OK, if sync list currently not works, we can set sync-stack as default false; set it to true in typespec-tests, and get things merged.
Default to true after mapPage fixed.

Too many potential conflicts in these tests.

**Pull request merged**